### PR TITLE
complete file handling for resources

### PR DESCRIPTION
**Pull request by ax-sc**
This PR aims to complete the file handling for resources by rendering and handling them with _vuejs_. As long as no other information is provided it should be handled as **work-in-progress**.

These known issues are still to be fixed:
- [X] ~~csrf_token missing on upload of new resource~~
- [X] ~~rendering of existing resources with _vuejs_~~
- [x] ~~if no name is present, the resource name can not be edited~~

These features are already implemented:
- [X] render resources with vuejs
- [X] make resource removable
- [X] make resource name editable

### CODE REVIEW

**Comment by ax-sc** [comment]
> These known issues are still to be fixed:
> - [ ] if no name is present, the resource name can not be edited

To handle this, i´m thinking about validating the form input on upload such that every resource must have a name. This is not fixing the mentioned issue completely, as we then also have to make sure that the name of the resource can not deleted by editing it afterwards.

@jnsdrtlf Do you think we should implement the input validation or add functionality to edit "empty" resource names as an alternative?

**Comment by 3j14** [comment]
Resource names should probably not be empty. Requiring names sounds beneficial. The `resource` Model already states `nullable=False`:

https://github.com/desklab/doku/blob/ee6e779260905af6afe3186ebc23cd67c3efca61/doku/models/resource.py#L17

**Comment by 3j14** [comment]
Simply adding the **`required`** attribute to the ´<intpu>´ element should do the trick

```html
<input type="text" name="name" required>
```

This check should also be implemented for the _POST_ request in flask.

File: doku/blueprints/api/v1/resource.py [hunk 1]
```diff
@@ -11,27 +11,52 @@
...
+    except ValidationError as e:
+        return jsonify(e.messages), BadRequest.code
+
+    db.session.commit()
+
+    result = schema.dump(resource)
+    return jsonify(result)
 
-@bp.route("/", methods=["POST"])
+@bp.route("/", methods=["PUT"])
```
Comment (3j14) [inline doku/blueprints/api/v1/resource.py]: `create` should be `POST` and `update` should be `PUT`, or am I mistaken?

**Review by 3j14 — CHANGES_REQUESTED** [review]
Besides the `POST` and `PUT` confusion this looks good to me. Regarding `POST` and `PUT`, have you checked the API requests in JavaScript?

**Comment by ax-sc** [comment]
> Besides the POST and PUT confusion this looks good to me. Regarding POST and PUT, have you checked the API requests in JavaScript?

I assume the API requests in JavaScript are correct as they use `put` in `updateResource` and `post` in `createResource`.

https://github.com/desklab/doku/blob/bd7d7e7c4b67656a6312e27c00e16ada1467bf1a/doku/static/src/api/resource.js#L6-L11

The confusion in `doku/blueprints/api/v1/resource.py ` is fixed in 0acfef9715a4f76d85c4922bc0fd8060e09888dd.

File: doku/blueprints/api/v1/resource.py [hunk 2]
```diff
@@ -11,27 +11,52 @@
...
+        unknown=EXCLUDE, session=db.session, instance=resource, partial=True
+    )
+
+    data = dict(request.form.copy())
+    if request.json is not None:
+        data.update(request.json)
+
+    if request.files.get("source", None) is not None:
+        file: FileStorage = request.files.get("source")
+        data["source"] = file.read()
```
Comment (3j14) [inline doku/blueprints/api/v1/resource.py]: This won’t work like this. For Stylesheets, we read the file and save it’s content in the database in the field _source_.

For resources, we need to save the actual file in the resource folder just like we do when creating the resource.

**Review by 3j14 — CHANGES_REQUESTED** [review]

File: doku/blueprints/api/v1/resource.py (hunk 2 above)
Comment (ax-sc) [reply doku/blueprints/api/v1/resource.py]: This method has not been used and was now removed in a9aa83b7fa88b329146bb701b7c7568376c83540.

**Review by 3j14 — APPROVED** [review]

**Pull request merged**